onmt/utils/loss.py

```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
     def _unbottle(self, _v, batch_size):
         return _v.view(-1, batch_size, _v.size(1))
 
+    def ignore_prompt(
+        self, batch, response_token="▁Response", response_right_pattern_length=3
+    ):
+        """
+        Mask the prompt in the target side of the bath examples in order
+            to set the loss of the prompt to zero.
+        For finetuning on specific tasks.
+        The masks are supposed to be properly handled by the loss criterion
+            (e.g. nn.CrossEntropyLoss )
+
+        Args:
+        batch: The current batch.
+        response_token (str): The token used the locate the end of the prompt
+        response_right_pattern_length (int). The number of tokens in the response
+            pattern after the response_token
+
+        Ex: The response pattern is '### Response : ｟newline｠',
+            the response_token is '▁Response'.
+            and the tokenized response pattern is '##', '#', '▁Response', '▁:', '▁', '<0x0A>'].
+            Then response_right_pattern_length = 3
+        """
+        import numpy as np
+
+        response_idx = self.vocab[response_token]  # 13291
+        ignore_index = self.padding_idx
+        nb_examples = batch["src"].size()[0]
+        for j in range(nb_examples):
+            # Locate the end of the prompt
+            response_start = np.where(
+                batch["src"][j, :, 0].cpu().numpy() == response_idx
+            )[0].tolist()[0]
+            response_start += response_right_pattern_length
```

Comment:
This `response_right_pattern_length` trick feels a bit weird to me (and probably hard to generalize, considering you would need to know how it's tokenized, etc.
Can't we make it work with just using some tokens? E.g. adding some kind of placeholder `｟_mask_before_｠` and put it where we would not need such trick.